src/lib/circuit_string.ts

```diff
@@ -32,38 +34,70 @@ export class Character extends CircuitValue {
   TODO: Add support for more character sets
   ASCII reference: https://en.wikipedia.org/wiki/ASCII#Printable_characters
   */
-  static check(value: Field): Boolean {
+  static check(value: Field) {
     console.log('check')
-    if (Circuit.inProver() || Circuit.inCheckedComputation()) {
-      const ret = Bool.and(
-        value.gte(Field(32)),
-        value.lte(Field(126))
-      );
+    value.assertGte(Field(32))
+    value.assertLte(Field(126))
+  }
+};
 
-      return ret.toBoolean();
-    }
+class NullCharacter extends Character {
+  constructor() {
+    super(Field(0));
+  }
 
-    return true;
+  static check(value: Field) {
+    value.assertEquals(Field(0))
   }
-};
+}
 
 export class CircuitString extends CircuitValue {
-  @prop values: Character[];
+  @prop usedLength: Field;
+  @prop maxLength: Field;
+
+
+  // arrayProp called later because we can't await isReady here
+  values;
 
   constructor(values: Character[]) {
     super();
 
-    this.values = values;
+    const usedLength = values.length;
+    const maxLength = DEFAULT_STRING_LENGTH;
+    values.length = maxLength;
+
+    // set the value of this.values to the input values, filled with null char
+    this.values = values.fill(new NullCharacter, usedLength);
+
+    this.usedLength = Field(usedLength);
```

Comment:
After seeing some of the problems in your circuits below, I think it's probably not helpful to store the `usedLength` on the class. Not storing it would force you to write your circuits in a way that they are agnostic to the actual length, instead using the static `maxLength`. This is the way to ensure that the usedLength actually stays dynamic, and not ends up being a constant that gets baked into the circuit